qiskit_dynamics/signals/__init__.py

```diff
@@ -12,14 +12,195 @@
+Signal objects represent a class of real-valued functions of the form:
+
+.. math::
+    s(t) = \textnormal{Re}[f(t)e^{i(2 \pi \nu t + \phi)}],
+
+where
+
+    * :math:`f` is a complex-valued function called the *envelope*,
+    * :math:`\nu \in \mathbb{R}` is the *carrier frequency*, and
+    * :math:`\phi \in \mathbb{R}` is the *phase*.
+
+There are two main signal classes:
+
+    #. :class:`~qiskit_dynamics.signals.Signal`: which
+       is instantiated with an *arbitrary* python callable function for the envelope, and
+    #. :class:`~qiskit_dynamics.signals.DiscreteSignal`: in which the envelope
+       is piecewise constant, specified as a list of samples with a sample width and start time.
+
+:class:`~qiskit_dynamics.signals.Signal` allows for complete generality,
+while :class:`~qiskit_dynamics.signals.DiscreteSignal` is internally implemented
+with array-based operations and geared more towards performance.
+
+
+Signal
+^^^^^^
+
+A :class:`~qiskit_dynamics.signals.Signal` is instantiated with an envelope.
+
+.. code-block:: python
+
+    f = lambda t: t**2
+    signal = Signal(envelope=f, carrier_freq=2.)
+
+**Important:** :class:`~qiskit_dynamics.signals.Signal` assumes the envelope ``f`` is
+*array vectorized* in the sense that ``f`` can operate on arrays of arbitrary shape
+and satisfy ``f(x)[idx] == f(x[idx])`` for a multidimensional index ``idx``. This
+can be ensured either by writing ``f`` to be vectorized, or by using the ``vectorize``
+function in ``numpy`` or ``jax.numpy``, e.g.:
+
+.. code-block:: python
+
+    import numpy as np
+    vectorized_f = np.vectorize(f)
+    signal = Signal(envelope=vectorized_f, carrier_freq=2.)
+
+Once instantiated, various functions associated with the siganl can be evaluated:
+
+.. code-block:: python
+
+    t = 0.5
+    # evaluate the full expression
+    signal(t)
+    # evaluate the envelope
+    signal.envelope(t)
+    # evaluate the complex value
+    signal.complex_value(t)
+
+We can also draw any of these functions over an interval:
+
+.. jupyter-execute::
+
+    from qiskit_dynamics.signals import Signal
+    signal = Signal(envelope=lambda t: t**2, carrier_freq=2.)
+
+    # draw with start point endpoint, and number of samples
+    signal.draw(t0=0., tf=2., n=100)
+
+
+Constant Signal
+^^^^^^^^^^^^^^^
+
+:class:`~qiskit_dynamics.signals.Signal` supports specification of a *constant signal*:
+
+.. code-block:: python
+
+    const = Signal(2.)
+
+This initializes the object to always return the constant ``2.``, and allows constants to be
+treated on the same footing as arbitrary :class:`~qiskit_dynamics.signals.Signal` instances.
+A :class:`~qiskit_dynamics.signals.Signal` operating in constant-mode can be checked via the
+boolean attribute ``const.is_constant``.
+
+
+DiscreteSignal
+^^^^^^^^^^^^^^
+
+A :class:`~qiskit_dynamics.signals.DiscreteSignal`  is specified
+as an array of samples with a sample width and an optional start point (defaults to ``0.``).
```

Comment:
I think this needs more description what sample width is (and should mention `dt`). Something like:

A discrete signal is specified by an Array [s_0, ..., s_k] and sample width `dt` and is evaluated as s(t) = s_{floor(t/dt)} (or something like that). By default discrete signals are defined to start at t=0, and hence have duration n * dt, where n is the number of samples, but a custom start time can be set by using the <whatever kwarg is> kwarg.

Reply:
Just put a version of this in.